### PR TITLE
TY&RES: Fully support trait bounds for complex types

### PR DESCRIPTION
**Pull request by vlad20012**
E.g. for associated types:
```rust
trait Foo {
    type Item;
    fn foo(&self) -> Self::Item;
}
trait Bar<A> { fn bar(&self); }
                //X
fn baz<T1, T2>(t: T1)
    where T1: Foo,
          T1::Item: Bar<T2> {
    t.foo().bar();
}         //^
```

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt [hunk 1]
```diff
@@ -1071,37 +1072,42 @@ class RsFnInferenceContext(
...
 
-        val impl = callee.source.impl
-        var typeParameters = if (impl != null) {
-            val typeParameters = instantiateBounds(impl)
-            impl.typeReference?.type?.substitute(typeParameters)?.let { ctx.combineTypes(callee.selfTy, it) }
-            if (callee.element.owner is RsAbstractableOwner.Trait) {
-                impl.traitRef?.resolveToBoundTrait?.substitute(typeParameters)?.subst ?: emptySubstitution
-            } else {
-                typeParameters
+        val source = callee.source
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt]: This is mostly refactoring

File: src/main/kotlin/org/rust/lang/core/types/ty/TyTypeParameter.kt [hunk 2]
```diff
@@ -32,6 +32,7 @@ class TyTypeParameter private constructor(
     override fun equals(other: Any?): Boolean = other is TyTypeParameter && other.parameter == parameter
     override fun hashCode(): Int = parameter.hashCode()
 
+    @Deprecated("Use ImplLookup.getEnvBoundTransitivelyFor")
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/types/ty/TyTypeParameter.kt]: I hope we'll refactor it in future (

**Comment by Undin** [comment]
@vlad20012 merge conflict

**Comment by vlad20012** [comment]
bors r=mchernyavsky

**Pull request merged**